src/crystal/system/unix/file_handle.cr

```diff
@@ -0,0 +1,172 @@
+require "c/fcntl"
+require "io/syscall"
+
+class Crystal::System::FileHandle
+  include IO::Syscall
+
+  @fd : Int32
+
+  @read_event : Event::Event?
+  @write_event : Event::Event?
+
+  @closed = false
+
+  def initialize(platform_specific : Int32)
+    @fd = platform_specific
+  end
+
+  def platform_specific : Int32
+    @fd
+  end
+
+  def read(slice : Bytes) : Int32
+    read_syscall_helper(slice, "Error reading file") do
+      # `to_i32` is acceptable because `Slice#size` is a Int32
+      LibC.read(@fd, slice, slice.size).to_i32
+    end
+  end
+
+  def write(slice : Bytes) : Nil
+    write_syscall_helper(slice, "Error writing file") do |slice|
+      LibC.write(@fd, slice, slice.size).tap do |return_code|
+        if return_code == -1 && Errno.value == Errno::EBADF
+          raise IO::Error.new "File not open for writing"
+        end
+      end
+    end
+  end
+
+  def closed? : Bool
+    @closed
+  end
+
+  def close : Nil
+    return if @closed
+
+    err = nil
+    if LibC.close(@fd) != 0
+      case Errno.value
+      when Errno::EINTR, Errno::EINPROGRESS
+        # ignore
+      else
+        err = Errno.new("Error closing file")
+      end
+    end
+
+    @closed = true
+
+    @read_event.try &.free
+    @read_event = nil
+
+    @write_event.try &.free
+    @write_event = nil
+
+    reschedule_waiting
+
+    raise err if err
+  end
+
+  def blocking? : Bool
+    (fcntl(LibC::F_GETFL) & LibC::O_NONBLOCK) == 0
+  end
+
+  def blocking=(value : Bool) : Bool
+    flags = fcntl(LibC::F_GETFL)
+    if value
+      flags &= ~LibC::O_NONBLOCK
+    else
+      flags |= LibC::O_NONBLOCK
+    end
+    fcntl(LibC::F_SETFL, flags)
+
+    value
+  end
+
+  def close_on_exec? : Bool
+    (fcntl(LibC::F_GETFD) & LibC::FD_CLOEXEC) == LibC::FD_CLOEXEC
+  end
+
+  def close_on_exec=(value : Bool) : Bool
+    flags = fcntl(LibC::F_GETFD)
+    if value
+      flags |= LibC::FD_CLOEXEC
+    else
+      flags &= ~LibC::FD_CLOEXEC
+    end
+    fcntl(LibC::F_SETFD, flags)
+
+    value
+  end
+
+  def seek(offset : Number, whence : IO::Seek = IO::Seek::Set) : Int64
+    check_open
+
+    seek_value = LibC.lseek(@fd, offset.to_i64, whence)
+
+    if seek_value == -1
+      raise Errno.new "Unable to seek"
+    end
+
+    seek_value.to_i64
+  end
+
+  def tty? : Bool
+    LibC.isatty(@fd) == 1
+  end
+
+  def reopen(other : FileHandle) : FileHandle
+    {% if LibC.methods.includes? "dup3".id %}
+      # dup doesn't copy the CLOEXEC flag, so copy it manually using dup3
+      flags = other.close_on_exec? ? LibC::O_CLOEXEC : 0
+      if LibC.dup3(other.platform_specific, self.platform_specific, flags) == -1
+        raise Errno.new("Could not reopen file descriptor")
+      end
+    {% else %}
+      # dup doesn't copy the CLOEXEC flag, copy it manually to the new
```

Comment:
Misplaced comment.

Reply:
I'm confused, it's missing a word but why is it misplaced? It's documenting the implementation so it should be inside the method?